Adding first draft of function for quantiles and qcut

kindofluke opened the pull request:
Not a very good JS programmer but I do use Zebras and quite like it. In this PR, I am trying to add `percentile` see [Numpy Percentile](https://numpy.org/doc/stable/reference/generated/numpy.percentile.html)

My eventual goal is to add `qcut` with functionality like [Pandas Qcut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html) but this is a prerequisite. 

Let me know what you think and if you have any guidance.

nickslevine commented:
Thanks! Will have a look ASAP.

nickslevine commented on source/percentile.js:
```diff
@@ -0,0 +1,56 @@
+    if (num <1 & num >= 0) {
+        return num
+    }
+    else if (num >=1 & num <= 100) {
```
Thanks for this! Definitely an important addition. 

I think it's worth deciding whether to require percentiles of the form 0<=p<=100 or 0<=p<=1.

The pandas quantile function does the latter, whereas the np percentiles function takes the former.

I don't have a view which one to choose, but handling both seems like it could cause confusion. Eg here, it's not obvious to the user whether p=1 is requesting the 1st %ile  or the 100th %ile. (In fact, if I'm reading this correctly, `p` becomes 0.01, but then `sortedArr[-1]` gets returned.)

nickslevine commented on source/percentile.js:
```diff
@@ -0,0 +1,56 @@
+ * @func
+ * @memberOf Z
+ * @category Analysis
+ * @param {Number} percentile to calculate
```
Suggest we state in the doc string something like "percentile to calculate in decimal form (eg 0.5 for 50th percentile, 0.75 for 75th percentile)" and update the below accordingly. to handle only one format (either integer or decimal).

nickslevine replied:
(see comment below)

kindofluke commented:
I am all for picking one and sticking with it. 

I guess I lean towards being between 0 and 1. I just think it's more explicit. 

If that is the case should the function be renamed to `quantile`? (this would be in line with pandas). 

Actually, I am going to do that because they make a pair `quantile` and `qcut` *quantile cut*. 


 

> On Feb 22, 2021, at 2:25 PM, nickslevine <notifications@github.com> wrote:
> 
> 
> @nickslevine requested changes on this pull request.
> 
> In source/percentile.js <https://github.com/nickslevine/zebras/pull/46#discussion_r580525278>:
> 
> > +    const i = p * length(sortedArr)
> +    var low_index = sortedArr[Math.floor(i - 1)]
> +    var high_index = sortedArr[Math.floor(i)]
> +    return (low_index + high_index)/2
> +
> +
> +    
> +})
> +
> +const diff = function(a, b) { return a - b; };
> +
> +function confirmPercentile(num) {
> +    if (num <1 & num >= 0) {
> +        return num
> +    }
> +    else if (num >=1 & num <= 100) {
> Thanks for this! Definitely an important addition.
> 
> I think it's worth deciding whether to require percentiles of the form 0<=p<=100 or 0<=p<=1.
> 
> The pandas quantile function does the latter, whereas the np percentiles function takes the former.
> 
> I don't have a view which one to choose, but handling both seems like it could cause confusion. Eg here, it's not obvious to the user whether p=1 is requesting the 1st %ile or the 100th %ile. (In fact, if I'm reading this correctly, p becomes 0.01, but then sortedArr[-1] gets returned.)
> 
> In source/percentile.js <https://github.com/nickslevine/zebras/pull/46#discussion_r580527287>:
> 
> > @@ -0,0 +1,56 @@
> +import { curry, sort, length, filter} from "ramda"
> +
> +import isNumeric from "./internal/isNumeric"
> +
> +/**
> + * Percentile of a value in a series of series using midpoint interpolation. see https://numpy.org/doc/stable/reference/generated/numpy.percentile.html
> + *
> + * @func
> + * @memberOf Z
> + * @category Analysis
> + * @param {Number} percentile to calculate
> Suggest we state in the doc string something like "percentile to calculate in decimal form (eg 0.5 for 50th percentile, 0.75 for 75th percentile)" and update the below accordingly. to handle only one format (either integer or decimal).
> 
> In source/percentile.js <https://github.com/nickslevine/zebras/pull/46#discussion_r580527487>:
> 
> > @@ -0,0 +1,56 @@
> +import { curry, sort, length, filter} from "ramda"
> +
> +import isNumeric from "./internal/isNumeric"
> +
> +/**
> + * Percentile of a value in a series of series using midpoint interpolation. see https://numpy.org/doc/stable/reference/generated/numpy.percentile.html
> + *
> + * @func
> + * @memberOf Z
> + * @category Analysis
> + * @param {Number} percentile to calculate
> (see comment below)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/nickslevine/zebras/pull/46#pullrequestreview-595700684>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AE6I5MFYCKO35FRL3F7KF3TTAKVSPANCNFSM4YBCEMUQ>.
>

nickslevine commented:
> I am all for picking one and sticking with it. I guess I lean towards being between 0 and 1. I just think it's more explicit. If that is the case should the function be renamed to `quantile`? (this would be in line with pandas). Actually, I am going to do that because they make a pair `quantile` and `qcut` *quantile cut*.
> […](#)

Yep makes sense - thanks!

kindofluke commented:
Updated with a bunch of changes: 

- [x] rename percentile to quantile
- [x] remove support for values 1 to 100 focus on supporting 0 to 1
- [x] added qcut
- [x] changed the errors to `RangeError` from `TypeError`
- [x] updated Typing
- [ ]  build and check documentation 

I don't think the tests for `qcut` are very good but I sort of got stumped on them a little. 

I'd love to close this one out and I am going to work on a new branch for to add `cut` and `hist`.

nickslevine commented:
Thanks! Looks good. Just merged.
